### PR TITLE
batch_norm: add output CB to UnpackToDestFp32 list for typecast path

### LINKED ISSUE #51315: batch_norm: output CB missing from UnpackToDestFp32 list on the fp32→bf16 typecast path
## Summary

`BatchNormFactory` omitted `output_tensor_cb` (`c_2`) from the `UnpackToDestFp32` set. On the
mixed-precision typecast path that CB holds **Float32** data *and* is unpacked back into DST, so
omitting it makes the unpacker route an fp32 staging tile through the non-fp32 path, degrading
output accuracy.

## Affected configuration

Requires **all** of:

1. `fp32_dest_acc_en == true` — the default (`batch_norm_utils.cpp` sets `default_fp32_acc = true`),
   so no explicit compute config is needed to hit this.
2. `needs_output_typecast == true`, i.e. `interm_data_format == Float32 && output != Float32`.
   Because output dtype is forced equal to input dtype, this means:
   **input/output `bfloat16` + at least one of `{batch_mean, batch_var, weight, bias}` `float32`** —
   a documented, supported mixed-precision configuration.

All-`bfloat16` and all-`float32` are both unaffected (in each, `needs_output_typecast` is false).

## Mechanism

In the typecast path the SFPU compute kernel uses `c_2` as an **fp32 staging buffer** and reads it
back for the dtype conversion:

- `batch_norm_sfpu_kernel.cpp:156` — `pack_tile(index * 2, cb_output_0)` writes the fp32 result
- `batch_norm_sfpu_kernel.cpp:171` — `copy_tile(cb_output_0, index, index * 2)` **unpacks it again**
- `:172-173` — `typecast_tile<Float32, Float16_b>` converts to the bf16 writer-facing CB (`c_9`)

`UnpackToDestFp32` is what makes the unpacker deliver full fp32 into DST. Without it on `c_2`, the
staging tile loses mantissa precision before the typecast even runs.

### How it was found
Surfaced as a Misc anomaly during the Metal 2.0 pre-port feasibility audit of normalization/batch_norm
(see METAL2_PREPORT_AUDIT.md in the op directory, anomaly 6): the two factories' UnpackToDestFp32
lists were inconsistent, and tracing which CBs are actually unpacked showed batch_norm's c_2
omission was the real gap.

### PR DESCRIPTION
**Pull request by VirdhatchaniKN**
### Issue
https://github.com/tenstorrent/tt-metal/issues/51315

## Summary

- Adds `output_tensor_cb` (`c_2`) to the `UnpackToDestFp32` list when
  `needs_output_typecast` is true in `BatchNormFactory`. On that path the SFPU
  kernel packs into `c_2` as fp32 staging and unpacks it via `copy_tile` before
  dtype conversion — host unpack config should match that.
- Adds `test_batch_norm_fp32_acc_output_typecast`:
  - bf16 I/O + fp32 mean/var (and optional weight/bias) → exercises
    `needs_output_typecast`
  - all-fp32 → control (no typecast)

Note: on WH with the current inputs, TT output is bit-identical with/without
this change, so the test is path coverage vs torch, not a numeric
pre-fix fail detector.

## Test plan

- [x] `test_batch_norm_fp32_acc_output_typecast` — 16 passed
- [x] `pytest tests/ttnn/unit_tests/operations/fused/test_batch_norm*.py`

### Pipeline status

- [x] Sanity
- [x] (Runtime) Sanity Test
- [x] BHPC

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:virdhatchani/BN_bug_port)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:virdhatchani/BN_bug_port)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-sanity-tests.yaml?query=branch:virdhatchani/BN_bug_port)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:virdhatchani/BN_bug_port)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:virdhatchani/BN_bug_port)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:virdhatchani/BN_bug_port)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:virdhatchani/BN_bug_port)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:virdhatchani/BN_bug_port)

### CODE REVIEW

File: tests/ttnn/unit_tests/operations/fused/test_batch_norm.py [hunk 1]
```diff
@@ -928,3 +928,90 @@ def to_tt(tensor):
...
+        input=in_ref,
+        running_mean=mean_ref,
+        running_var=var_ref,
+        weight=weight_ref,
+        bias=bias_ref,
+    ).to(tt_output.dtype)
+
+    if needs_output_typecast:
+        assert_numeric_metrics(
+            torch_result, tt_output, pcc_threshold=0.999, rtol=1e-2, atol=0.1, frobenius_threshold=0.02
```
Comment (Aswinmcw) [inline tests/ttnn/unit_tests/operations/fused/test_batch_norm.py:1012]: Non-blocking: worth confirming this typecast-case assertion actually fails on pre-fix main. Input range is small (5-10) and pcc 0.999 is fairly loose for a mantissa-precision regression — if the buggy (non-fp32-unpack) output still clears these thresholds, the test passes both with and without the fix and doesn't guard the regression. If you've verified it fails without the change, ignore this.

**Review by Aswinmcw — APPROVED** [review]
LGTM — fix is correctly gated on fp32_dest_acc_en + needs_output_typecast and targets exactly c_2.

File: tests/ttnn/unit_tests/operations/fused/test_batch_norm.py (hunk 1 above)
Comment (VirdhatchaniKN) [reply tests/ttnn/unit_tests/operations/fused/test_batch_norm.py:1012]: Checked on WH — you're right. With the factory fix reverted, the typecast case still passes, and TT output is bit-identical to the fixed build (no elems differ). So this test isn't a numeric regression detector with the current inputs; it's mixed-precision path coverage for needs_output_typecast. Keeping the host fix so c_2 matches the kernel unpack on that path. Updated the PR wording accordingly.

**Review by bbradelTT — APPROVED** [review]
